src/fairseq2/metrics/text/_bleu.py

```diff
@@ -65,14 +76,16 @@ def update(self, refs: Sequence[str], hyps: Sequence[str]) -> Self:
         :param refs: The references.
         :param hyps: The hypotheses.
         """
-        bleu = corpus_bleu(hyps, [refs], tokenize=self.tokenizer)
+        bleu = self._metric_class.corpus_score(hyps, [refs])
 
         self.sys_len += bleu.sys_len
         self.ref_len += bleu.ref_len
 
-        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device)
-        self.total_ngrams += torch.tensor(bleu.totals, device=self.device)
-
+        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device).clone()
+        self.total_ngrams += torch.tensor(bleu.totals, device=self.device).clone()
+        self._steps += 1
+        if self._steps % 500 == 0:
```

Comment:
Why do we still need to call GC here if we have fixed the memory leak? Having said that, we call GC every 1000 step in the trainer (which is configurable here: https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/recipes/config.py#L171-L172)

I think a metric implementation should not have this kind of "hidden" behavior that might impact the overall job throughput. Depending on the memory allocation pattern of the job, a GC can become quite expensive.

Reply:
it runs ok without it

Reply:
@cbalioglu : does `gc_every_n_steps` acts also for validation steps ?

Reply:
actually, no - it's better with an internal gc.collect() ...

Reply:
We do a CUDA cache allocator clean-up, but not GC: https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/recipes/_trainer.py#L1167-L1168

We can also do GC there though.

> actually, no - it's better with an internal gc.collect()

Do you mean you still get a leak when you don't call gc.collect() even after the refactoring of the corpus call?